### ISSUE TITLE
SortMergeJoin is not thread-safe

### ISSUE THREAD

**Issue by mrks** [hyrise/hyrise#1643]
`[ FAILED ] SQLiteTestRunnerUnencoded/SQLiteTestRunner.CompareToSQLite/106, where GetParam() = ("SELECT * FROM id_int_int_int_100 AS t1 LEFT JOIN id_int_int_int_100 AS t2 ON t1.a >= t2.a AND t1.b >= t2.b AND t1.c >= t2.c;", false, Unencoded) (106 ms)`

Running `./hyriseTest --gtest_filter=SQLiteTestRunnerUnencoded/SQLiteTestRunner.CompareToSQLite/106` either crashes, passes, or fails on clang-debug with fed77246c3751d2d50566fb27f7898b0c6b3db15 (download zip from Github).

Also, the JoinTestRunner should have a multi-threaded configuration.

**Comment by mrks** [comment]
These rows show up in our result but not in sqlite (which has no NULLs):

```
          id        a        b        c       id        a        b        c
         int      int      int      int      int      int      int      int
 476      16       41       10       30     NULL     NULL     NULL     NULL
 477      84       72       19        2     NULL     NULL     NULL     NULL
```

@mrzzzrm Does this look familiar in any way by chance?

**Comment by mrks** [comment]
This never worked:

```
git submodule update
echo 'SELECT * FROM id_int_int_int_100 AS t1 LEFT JOIN id_int_int_int_100 AS t2 ON t1.a >= t2.a AND t1.b >= t2.b AND t1.c >= t2.c;' > resources/test_data/sqlite_testrunner_queries.sql
cd build
ninja hyriseTest
timeout -s SIGKILL --preserve-status 120 ./hyriseTest --gtest_filter=SQLiteTestRunnerUnencoded/SQLiteTestRunner.CompareToSQLite/0 --gtest_repeat=20
ret=$?
cd ..
git checkout resources/test_data/sqlite_testrunner_queries.sql
if [ "$ret" -eq "0" ]; then return 0; fi; return 1
```

```
95561eca450cac55abc696e0c74be01bd68998ac is the first bad commit
commit 95561eca450cac55abc696e0c74be01bd68998ac
Author: Kolpa <kolyaopahle@gmail.com>
Date:   Tue Mar 26 11:51:15 2019 +0100

    [DYOD18/19] Multi Predicate Join Support (#1482)
```

**Comment by mrks** [comment]
Here's a different Hyrise result:

```
 32       11       59       19     NULL     NULL     NULL     NULL
 32       11       59       19       32       11       59       19
 32       11       59       19       43       11        1       19
```

I find it interesting that the (wrong) NULL row comes first. Shouldn't the non-matches be written at the end?

**Comment by mrzzzrm** [comment]
Nice, no, doesn't look familiar. Happy drilling ;)
You could checkout the JoinTestRunner branch and make the query plans use the JoinVerificationOperator (by altering the LQPTranslator) instead of JoinHash, JSM, etc.

**Comment by mrks** [comment]
`_left_row_id_has_match` is written happily from multiple threads. I'm surprised that tsan isn't red-faced.

Is there any reason why that and _right_row_id_has_match are shared across different clusters?

I won't find the time to fix this in the next days. Up for grabs.

**Comment by Bouncner** [comment]
Cannot look into the code right now, but there might be cases where matches can come from other clusters (e.g., for `<` joins).

**Comment by mrzzzrm** [comment]
any idea why this would only fail if the query is run in isolation, but never if the entire testsuite is run?

**Comment by mrks** [comment]
"Never" is not correct. What is true is that it happens significantly more often when the test is isolated. No idea yet why that might be the case.

**Comment by Bouncner** [comment]
Can't say for sure that 106 is no longer a problem after #2315, but at least this test does not crash.

What crashes with an enabled scheduler is the following query:
```
Note: Google Test filter = *Line391WithUnencoded*
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from SQLiteTestRunnerUnencoded/SQLiteTestRunner
[ RUN      ] SQLiteTestRunnerUnencoded/SQLiteTestRunner.CompareToSQLite/Line391WithUnencoded
1
0x7fa33120fed8
libc++abi.dylib: terminating with uncaught exception of type std::logic_error: src/lib/operators/abstract_operator.cpp:72 Operator is already being executed.
[1]    12410 abort      ./hyriseTest --gtest_filter="*Line391WithUnencoded*"
```

Play around branch:
martin/#1643. Looks like the aggregate in the subquery is already executed?

**Comment by mrks** [comment]
Pinging @julianmenzler who touched this recently.

**Comment by Bouncner** [comment]
Fixed by #2315.
